### ISSUE TITLE
Performance issues using Hamming distance in tree-based estimators

### ISSUE THREAD

**Issue by grovduck** [lemma-osu/sknnr#100]
In #85, we implemented the `RFNNRegressor` to mimic similar functionality in `yaImpute` and in #98, we introduced the `WeightedHammingDistanceMixin` to serve as the `kneighbors` distance metric which overcame some numeric instability found in `scipy.spatial.distance.hamming` owing to the use of `np.dot` in the function (detailed in [this thread](https://github.com/lemma-osu/sknnr/pull/98#issuecomment-3207781962)).  

When experimenting with using `RFNNRegressor` in a spatial context using `sklearn-raster`, we noticed very slow runtimes even over small spatial domains (e.g. 50 x 50 windows) and relatively small number of forests and trees used to create the `RFNNRegressor` estimator.  Through profiling (with both `cProfile` and `line_profiler`), the bottleneck appears to be the element-wise calculation of Hamming distance through `kneighbors`.  For example, a 50 x 50 window with 2000 samples across 100 estimators (trees) in a single forest has to do (2500 * 2000 ~= 4.3 million) python calls through `sklearn`.

Ideally, fewer Python calls for distance finding would be better.  For a single query point `q`, we can create the node dissimilarity matrix with the reference node IDs `R` with a single numpy call (i.e. `diff = R != q`).  This code sample illustrates the current implementation (using a single `RandomForestClassifier`) versus a much faster vectorized approach.
```
# In ipython

from sklearn.datasets import make_classification
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsRegressor
from scipy.spatial.distance import hamming

X, y = make_classification(
    n_samples=2000,
    n_features=10,
    n_informative=6,
    n_redundant=0,
    random_state=0,
    shuffle=False
)
clf = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=0)
clf = clf.fit(X, y)

# Create the reference node matrix (R)
R = clf.apply(X)

# Get a sample query point (q)
q = clf.apply(X[0].reshape(1, -1))

# Current method using a kNN regressor
# Note that I'm using `scipy.spatial.distance.hamming` as a proxy for our
# custom `WeightedHammingDistanceMetric` as precision isn't critical for
# this test
knn_est = KNeighborsRegressor(n_neighbors=5, metric="hamming").fit(R, y)

# Get timing on getting the top 5 neighbors/distances
# 1.72 ms ± 85.6 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit knn_est.kneighbors(q)

# New method comparing q against all R
# 9.13 ns ± 0.624 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
w = np.ones(R.shape[1], dtype=np.int32)
%timeit "diff = R != q; d = diff @ w; np.argpartition(d, 5)[:5]"

# Assert that the same neighbors are identified
old = knn_est.kneighbors(q)[:5][1]

diff = R != q
d = diff @ w
new = np.argpartition(d, 5)[:5]
print(np.all(old == new))
# True
```
This represents a massive (200,000x) speedup if I'm not doing something stupid.  

@aazuspan, I haven't totally thought this through, but what I think this means is that `RFNNRegressor` (and other tree-based estimators) at the very least need to override `TransformedKNeighborsRegressor.kneighbors` and I'm not sure there would be value continuing to have `TransformedKNNRegressor` as a base class.  Other than this small example, I haven't yet implemented this, but I don't foresee issues handling individual tree weights (already demonstrated above) or forest weights.  If you have other thoughts on this, I'd be grateful to hear them.

PS. This was the source of what I was describing last week and why you wouldn't have seen the same slowdowns just using RF and GB directly (i.e. without using `.apply()`)

**Comment by aazuspan** [comment]
>  For example, a 50 x 50 window with 2000 samples across 100 estimators (trees) in a single forest has to do (2500 * 2000 ~= 4.3 million) python calls through sklearn.

Yikes! So `sklearn` is doing Python loops over samples when calculating `kneighbors`? Is there something specific about RF-NN that's forcing that slow implementation, or is it just the high dimensionality and/or using Hamming distance that's highlighting some inherent limitations?

> This code sample illustrates the current implementation (using a single RandomForestClassifier) versus a much faster vectorized approach.

Sorry if this is a red herring, but I stumbled on a weird timing discrepancy trying to run this and wanted to make sure I'm not doing something wrong before diving deeper. When I copied your code, I moved the inline `%timeit` calls into cell-level `%%timeit` so that I could split out the string expression and make sure I was reading it correctly, i.e.

```python
%%timeit

w = np.ones(R.shape[1], dtype=np.int32)
diff = R != q
d = diff @ w
np.argpartition(d, 5)[:5]
# 958 μs ± 83.6 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

As you can see, that ran in about ~1ms - still almost twice the speed of the `sklearn` version (1.69 ms for me), but nowhere near the original speedup for some reason. I went back and ran it in the string form as you originally wrote it and that *does* give me similar timing to yours (13.6 ns). *But*, if I run the `sklearn` code as a string, that also runs in a tiny fraction of the non-string time:

```python
%timeit "knn_est.kneighbors(q)"
# 14 ns ± 0.216 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
```

I've never encountered this before and don't have any idea what's going on, but is it possible this is a weird fluke with `timeit` that's throwing off our results and breaking the comparison?

**Comment by aazuspan** [comment]
> As you caught, when putting quotation marks around the inlined version of %timeit, I think it's testing the speed of creating a string and not evaluating the set of commands. Obviously something I hadn't run into before!

Tricky - I would have assumed that should work like the `timeit` module where strings are parsed and evaluated!

> I think we can leave the creation of the w array outside the timing because that should be need to be done only once for all targets.

I was wondering about that. Initializing once outside the timing loop makes perfect sense.

>  For our current implementation, "auto" gets changed to "brute" which sets neighbor finding to be pairwise distance finding.

Excuse what's probably a dumb question because I have no idea how those algorithms work, but I assume we can't just override "auto" to force one of the "fast" algorithms? Presumably they're incompatible and/or wouldn't be any more efficient for our case?

> EDIT: Probably a better option is to keep metric="hamming" on initialization of KNeighborsRegressor, get the weights from fitting the RF/GB trees and then using est.set_params(metric_params={"w": w}).

That definitely sounds like a promising option. Once `metric_params` are set on the regressor, they're automatically picked up and used by the metric? I imagine using `"hamming"` re-introduces the `np.dot` imprecision issue for regression testing that we dealt with in #98, but that seems like something we can work around. I was worried that we'd considered this approach in that PR and rejected it due to some tricky edge case, but on a quick skim I didn't see where we'd talked about combining the built-in `"hamming"` metric with `metric_params`.
